### PR TITLE
Support covariant viewdatadictionaries

### PR DESCRIPTION
**Pull request by javiercn**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs [hunk 1]
```diff
@@ -63,6 +63,28 @@ public override void Contextualize(ViewContext viewContext)
...
+            {
+                // The view data that we have at this point might be of a more derived type than the one defined at compile time.
+                // For example ViewDataDictionary<Cat> where our TModel is animal and Cat : Animal.
+                // This can happen if someone modified the model type through the page application model.
+                // In that case, we check if the type of the current view data, 'ViewData<TRuntime>' is "covariant" with the
+                // one defined at compile time 'ViewData<TModel>'
+                // When a type T<K> is covariant it means that an instance of that type T<'a> can be safely assigned to a variable
+                // of type T<'b> when 'a is of a more derived type of 'b. An example of this would be the array type, in which an
+                // instance of an array int[] can be safely assigned to a variable of type object[]. Or in code,
+                // var int [] integers = new[]{ 1, 2, 3, 4};
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs]:
```C#
Just use VDD as an example. 

a = VDD<Derived>
b = VDD<Base>

b = a;
```

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs [hunk 2]
```diff
@@ -76,6 +98,20 @@ public override void Contextualize(ViewContext viewContext)
...
 
+        private static Type GetViewDataOfTRuntimeType(ViewContext viewContext)
+        {
+            var viewDataType = viewContext.ViewData.GetType();
+            var viewDataArguments = viewDataType.GetGenericArguments();
+            if (viewDataArguments.Length != 1)
+            {
+                return null;
+            }
+            else
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs]: the `else` isn't necessary

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs [hunk 3]
```diff
@@ -63,6 +63,28 @@ public override void Contextualize(ViewContext viewContext)
...
             ViewData = viewContext.ViewData as ViewDataDictionary<TModel>;
+
+            if (ViewData == null)
+            {
+                // The view data that we have at this point might be of a more derived type than the one defined at compile time.
+                // For example ViewDataDictionary<Cat> where our TModel is animal and Cat : Animal.
+                // This can happen if someone modified the model type through the page application model.
+                // In that case, we check if the type of the current view data, 'ViewData<TRuntime>' is "covariant" with the
+                // one defined at compile time 'ViewData<TModel>'
+                // When a type T<K> is covariant it means that an instance of that type T<'a> can be safely assigned to a variable
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs]: Super don't need this explanation here. Just say we're allowing this case `VDD<Base> base; VDD<Derived> derivied; where Derived inherits from Base. We'll allow this case base = derived;

Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs]: Agree we don't need all the words explaining covariant because that's what Bing is for. However, *do* need to explain that `ViewDataDictionary<Derived>` can't occur for regular views -- only pages.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs (hunk 1 above)
Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs]: Even better, stick with `ViewDataDictionary<Cat>`

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs [hunk 4]
```diff
@@ -63,6 +63,28 @@ public override void Contextualize(ViewContext viewContext)
...
+                // of type T<'b> when 'a is of a more derived type of 'b. An example of this would be the array type, in which an
+                // instance of an array int[] can be safely assigned to a variable of type object[]. Or in code,
+                // var int [] integers = new[]{ 1, 2, 3, 4};
+                // var object [] data = integers;
+                // The statement above would not be possible in the general case given the fact that int[] doesn't extend object[], but
+                // its valid because arrays are covariant and int inherits from object.
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs]: `it's`

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs [hunk 5]
```diff
@@ -63,6 +63,28 @@ public override void Contextualize(ViewContext viewContext)
...
+                var runtimeType = GetViewDataOfTRuntimeType(viewContext);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelperOfT.cs]: This helper is doing scary runtime Reflection. But, it should just return `viewContext.ViewData.ModelMetadata.ModelType`. Use that instead and confirm your tests still work.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/DefaultTemplatesUtilities.cs [hunk 6]
```diff
@@ -42,6 +42,10 @@ public ObjectTemplateModel()
             public object ComplexInnerModel { get; set; }
         }
 
+        public class DerivedObjectTemplateModel : ObjectTemplateModel
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/DefaultTemplatesUtilities.cs]: Why is this here instead of `HtmlHelperTest`?

Comment (javiercn) [reply test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/DefaultTemplatesUtilities.cs]: Because other tests in HtmlHelperTests were using these models.

**Comment by javiercn** [comment]
Addressed feedback and updated @pranavkm @dougbu

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/DefaultTemplatesUtilities.cs (hunk 6 above)
Comment (dougbu) [reply test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/DefaultTemplatesUtilities.cs]: Please move this particular one to `HtmlHelperTest`. The ones here are used in multiple test classes.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTest.cs [hunk 7]
```diff
@@ -286,6 +287,58 @@ public void RawString_ReturnsExpectedString(string value, string expectedString)
             Assert.Equal(expectedString, result.ToString());
         }
 
+        [Fact]
+        public void Contextualize_WorksWithViewDataDictionariesOfTypesSpecializingTheModelType()
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTest.cs]: Stick w/ one terminology: `Contextualize_WorksWithCovariantViewDataDictionary`

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTest.cs [hunk 8]
```diff
@@ -286,6 +287,58 @@ public void RawString_ReturnsExpectedString(string value, string expectedString)
...
+            // Arrange
+            var helperToContextualize = DefaultTemplatesUtilities
+                .GetHtmlHelper<DefaultTemplatesUtilities.ObjectTemplateModel>(model: null);
+
+            var viewContext = DefaultTemplatesUtilities
+                .GetHtmlHelper<DefaultTemplatesUtilities.ObjectWithScaffoldColumn>(model: null)
+                .ViewContext;
+
+            // Act & Assert
+            Assert.Throws<ArgumentException>(() => helperToContextualize.Contextualize(viewContext));
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTest.cs]: Check the message. Want to be sure it is what we expect to be thrown.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTest.cs [hunk 9]
```diff
@@ -286,6 +287,58 @@ public void RawString_ReturnsExpectedString(string value, string expectedString)
...
+            var helperToContextualize = DefaultTemplatesUtilities
+                .GetHtmlHelper<DefaultTemplatesUtilities.ObjectTemplateModel>(model: null);
+
+            var viewContext = DefaultTemplatesUtilities
+                .GetHtmlHelper<DefaultTemplatesUtilities.ObjectWithScaffoldColumn>(model: null)
+                .ViewContext;
+            viewContext.ViewData = new ViewDataDictionary(viewContext.ViewData);
+
+            // Act & Assert
+            Assert.Throws<ArgumentException>(() => helperToContextualize.Contextualize(viewContext));
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTest.cs]: Check the message, should look similar to error in above test.
